src/v/archival/service.cc

```diff
@@ -336,10 +339,10 @@ ss::future<> scheduler_service_impl::reconcile_archivers() {
     auto snapshot = lm.get_all_ntps();
     std::vector<model::ntp> to_remove;
     std::vector<model::ntp> to_create;
-    // find ntps that exist in _archivers but no longer present in log_manager
+    // find archivers that have already stopped
     for (const auto& [ntp, archiver] : _archivers) {
         auto p = pm.get(ntp);
-        if (!snapshot.contains(ntp) || !p || !p->is_leader()) {
+        if (!snapshot.contains(ntp) || !p || archiver->upload_loop_stopped()) {
```

Comment:
I think the following situation is possible
- the term of the ntp get incremented but the leadership stays on the same node
- in this case we will remove the old archiver on next iteration of the reconciliation loop but we won't add the new one, for the new term
- the new archiver with the increased term will be added on the next iteration after that
- so if the reconciliation loop is started every 10s, we will start uploading data after 20s

maybe instead of that we can use ntp+term-id as a key in the map? in this case we will create new archiver on the same iteration

Reply:
I don't like ntp+term_id as the key because it means possibly increased archiver concurrency, but this is exactly the thing that we are trying to avoid :)

To mitigate this problem we can
1) decrease the sleep period to say 1s. This is on par with typical leader election time and (I hope) won't result in any significant additional load.
2) stop then start archivers sequentially. Not sure if this is better

Reply:
I don't think that we should do option 2. but option 1. should work fine.
With option 2 we may eventually see some overlap/concurrency between different loop iterations.